### PR TITLE
CI: Set up CI for the free-threaded build of 3.13

### PR DESCRIPTION
**Pull request by lysnikolaou**
Related to #59057.

### CODE REVIEW

**Comment by lysnikolaou** [comment]
Good news is free-threading CI is passing! 🎉 

Bad news is there's failures in `Docstring validation, typing, and other manual pre-commit hooks` and `windows-latest actions-312.yaml`, but I don't know how these can be related to this PR (they also appear to be passing on latest `main`).

**Comment by lysnikolaou** [comment]
Both are unrelated to this PR. Windows was flaky and got green after a rerun. Opened a new issue about the doctest failures in #59093.

File: .github/workflows/unit-tests.yml [hunk 1]
```diff
@@ -346,6 +346,49 @@ jobs:
...
+    defaults:
+      run:
+        shell: bash -eou pipefail {0}
+    runs-on: ubuntu-22.04
+
+    timeout-minutes: 90
+
+    concurrency:
+      # https://github.community/t/concurrecy-not-work-for-push/183068/7
+      group: ${{ github.event_name == 'push' && github.run_number || github.ref }}-${{ matrix.os }}-${{ matrix.pytest_target }}-dev
```
Comment (mroeschke) [inline .github/workflows/unit-tests.yml:359]:
```suggestion
      group: ${{ github.event_name == 'push' && github.run_number || github.ref }}-${{ matrix.os }}-python-freethreading-dev
```

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @lysnikolaou

**Pull request merged**